### PR TITLE
feat(diag): suggest "did you mean: X?" for undefined identifiers

### PR DESCRIPTION
**Pull request by liebe-magi**
## Summary

First sub-PR of v0.4.1's B4 (Better diagnostics) work. When a script references a variable or calls a function that doesn't resolve in scope, the runtime now appends a Levenshtein-based hint to the standard error:

\`\`\`
forge counter: arcana = 10;
unveil(countar);

# before: Error at line 2, column 8: Variable countar is not defined!
# after:  Error at line 2, column 8: Variable countar (did you mean: counter?) is not defined!
\`\`\`

Same enrichment for function calls:

\`\`\`
unveels("oops");
# before: Variable unveels is not defined!
# after:  Variable unveels (did you mean: unveil?) is not defined!
\`\`\`

When no candidate is within edit-distance the message is unchanged from the pre-suggestion era — truly unknown identifiers still surface as bare \`Variable X is not defined!\`. Multiple plausible candidates render as \`(did you mean: X, or Y?)\`.

## Why this approach

The user-visible shape of the message is enriched **without changing the public \`EvalError\` API**. \`EvalError::UndefinedVariable(String, Option<LineInfo>)\` keeps its existing tuple shape; the suggestion is folded into the variable-name string at the error-construction sites. Pattern-match downstream code keeps working, no \`#[non_exhaustive]\` needed yet (saved for PR4-C).

## Implementation

- New \`crates/abyss-interpreter/src/diagnostics.rs\` module with:
  - \`levenshtein(a, b) -> usize\` — Wagner–Fischer DP on \`Vec<char>\`, O(len(a) * len(b)) time, O(len(b)) space.
  - \`did_you_mean(target, candidates, max_count) -> Vec<String>\` — scores all candidates, filters to those within a length-scaled distance threshold (\`max(2, len/3)\` capped at \`len/2\`), sorts ascending, dedupes, truncates.
  - \`label_with_suggestions(name, suggestions) -> String\` — renders \`name (did you mean: X?)\` for one or \`name (did you mean: X, or Y?)\` for many; passes through bare \`name\` on empty.
- Two helpers on \`RuntimeEnv\`:
  - \`undefined_variable_error(name, line_info)\` searches the variable scopes.
  - \`undefined_function_error(name, line_info)\` searches the function scopes.
  Both build \`EvalError::UndefinedVariable\` with the enriched label.
- All seven internal \`EvalError::UndefinedVariable\` raising sites route through these helpers — variable lookups (\`AST::Var\`, method-receiver base name, index/field assignment targets, \`update_var\`) use the variable helper; the function-call site uses the function helper. Test sites and the \`Display\` impl are unchanged.

## Verification

- [x] 9 unit tests in \`diagnostics\`: Levenshtein edge cases (empty inputs, \`kitten\`/\`sitting\` reference), \`did_you_mean\` ordering / dedup / max_count, label formatting for 0/1/N suggestions.
- [x] Full integration suite (~287 tests) continues to pass — existing assertions on \`EvalError::UndefinedVariable(name, _)\` stay green because the helper passes through the bare name when no suggestions are available.
- [x] \`cargo fmt --all -- --check\` clean.
- [x] \`cargo clippy --all-targets --all-features -- -D warnings\` clean.
- [x] Live release-binary verification covered four behavioural shapes:
  - typo'd variable → suggests the close match.
  - typo'd function → suggests the close function name.
  - no plausible match (\`zzzzzzzzzzz\`) → falls back to bare error.
  - multiple plausible matches → renders comma-separated with \`or\`.

## Limitations / known caveats

- The function-call error path still labels its result \`"Variable …"\` (because it reuses the existing \`EvalError::UndefinedVariable\` variant). That's a pre-existing inaccuracy; renaming the variant would itself be a breaking change. PR4-C will address the broader runtime-error overhaul where this can be revisited cleanly.
- Suggestions don't yet kick in for \`Method X is not defined for scroll\` or \`Field X is not defined on artifact Player\`. Those are PR4-B.

## Out of scope (sequenced for the next two B4 sub-PRs)

- **PR4-B**: list available methods / fields on type-mismatched receivers.
- **PR4-C**: migrate runtime errors to \`ariadne\` reporting (multi-span, \`expected vs got\`) and add \`#[non_exhaustive]\` to \`EvalError\` so future variant additions stay non-breaking.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/abyss-interpreter/src/eval/statements.rs [hunk 1]
```diff
@@ -259,9 +256,11 @@ pub fn evaluate(ast: &AST, env: &mut RuntimeEnv) -> Result<EvalResult, EvalError
...
             let new_value = eval_result_to_value_checked(evaluate(value, env)?, line_info.clone())?;
 
-            let var_info = env.get_var_mut(&base_name).ok_or_else(|| {
-                EvalError::UndefinedVariable(base_name.clone(), line_info.clone())
-            })?;
+            // Look up before borrowing mutably so the error builder can read scope.
+            if env.get_var(&base_name).is_none() {
+                return Err(env.undefined_variable_error(&base_name, line_info.clone()));
+            }
+            let var_info = env.get_var_mut(&base_name).expect("just confirmed present");
```
[the remark this answers is not in the document]
Comment (liebe-magi) [reply crates/abyss-interpreter/src/eval/statements.rs]: Adopted in commit bf180a8 — replaced the pre-check + `expect` with a single `match env.get_var_mut(...)` whose `None` arm returns the helper's error directly. NLL handles the borrow correctly (the &mut from `get_var_mut` is only held in the `Some` arm, so the &self call to `undefined_variable_error` in `None` is fine). Eliminates both the double lookup and the `expect` panic path.

File: crates/abyss-interpreter/src/eval/statements.rs [hunk 2]
```diff
@@ -317,9 +316,11 @@ pub fn evaluate(ast: &AST, env: &mut RuntimeEnv) -> Result<EvalResult, EvalError
...
             let evaluated_value = evaluate(value, env)?;
 
-            let var_info = env.get_var_mut(&base_name).ok_or_else(|| {
-                EvalError::UndefinedVariable(base_name.clone(), line_info.clone())
-            })?;
+            // Look up before borrowing mutably so the error builder can read scope.
+            if env.get_var(&base_name).is_none() {
+                return Err(env.undefined_variable_error(&base_name, line_info.clone()));
+            }
+            let var_info = env.get_var_mut(&base_name).expect("just confirmed present");
```
[the remark this answers is not in the document]
Comment (liebe-magi) [reply crates/abyss-interpreter/src/eval/statements.rs]: Same fix applied alongside the previous one in commit bf180a8.

File: crates/abyss-interpreter/src/diagnostics.rs [hunk 3]
```diff
@@ -0,0 +1,146 @@
+//! Lightweight helpers shared between error-reporting sites.
+//!
+//! The interpreter raises errors as `EvalError` strings; this module supplies
+//! the bits of formatting reused across several call sites — currently the
+//! Levenshtein-based "did you mean?" suggestion logic that enriches messages
+//! about undefined variables, functions, methods, and artifact fields.
```
[the remark this answers is not in the document]
Comment (liebe-magi) [reply crates/abyss-interpreter/src/diagnostics.rs]: Good catch — over-promised in the doc. Trimmed to describe what's actually shipped (variables and functions) with a short note that the same primitives will be reused for methods/fields when PR4-B lands. Fixed in commit bf180a8.

File: crates/abyss-interpreter/src/diagnostics.rs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    let max_distance = ((target_len / 3).max(2)).min(target_len.div_ceil(2).max(1));
+
+    let mut scored: Vec<(usize, String)> = candidates
+        .into_iter()
+        .filter(|name| *name != target)
+        .map(|name| (levenshtein(target, name), name.to_string()))
+        .filter(|(dist, _)| *dist <= max_distance)
+        .collect();
+    scored.sort_by_key(|(dist, _)| *dist);
+    scored.dedup_by(|a, b| a.1 == b.1);
```
[the remark this answers is not in the document]
Comment (liebe-magi) [reply crates/abyss-interpreter/src/diagnostics.rs]: Real concern — `HashMap::keys()` was the actual candidate iterator at the call sites, so the unstable sort could surface different orderings between runs. Switched to `sort_by(|a, b| a.0.cmp(&b.0).then_with(|| a.1.cmp(&b.1)))` so the result is byte-for-byte identical regardless of input order. Locked it in with a new test `did_you_mean_breaks_ties_alphabetically_for_determinism` that exercises both input orderings. Fixed in bf180a8.

File: crates/abyss-interpreter/src/diagnostics.rs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+///
+/// When the suggestion list is empty the original `name` is returned
+/// unchanged, so call sites can hand the result straight to
+/// `EvalError::UndefinedVariable(...)` without further branching.
+pub(crate) fn label_with_suggestions(name: &str, suggestions: &[String]) -> String {
+    match suggestions {
+        [] => name.to_string(),
+        [only] => format!("{} (did you mean: {}?)", name, only),
+        many => format!("{} (did you mean: {}?)", name, many.join(", or ")),
+    }
```
[the remark this answers is not in the document]
Comment (liebe-magi) [reply crates/abyss-interpreter/src/diagnostics.rs:96]: Adopted in bf180a8. The formatter now special-cases the four arms cleanly:

- 1 → `X`
- 2 → `X or Y` (no comma — natural for two)
- 3+ → `X, Y, or Z` (Oxford comma + final `or`)

The previous test `label_with_suggestions_formats_multiple_matches` was renamed to `_formats_two_matches` and a new `_formats_three_or_more_matches_with_oxford_comma` test was added so both shapes are locked in.
